codebasin/config.py

```diff
@@ -270,6 +306,225 @@ def load_codebase(config, rootdir):
     return codebase
 
 
+_importcfg = None
+def load_importcfg():
+    """
+    Load the import configuration file, if it exists.
+    """
+    global _importcfg
+    _importcfg = collections.defaultdict(list)
+    path = ".cbi/importcfg.json"
+    if os.path.exists(path):
+        log.info(f"Found import configuration file at {path}")
+        with open(path, "r") as f:
+            try:
+                _importcfg_json = json.load(f)
+                for compiler in _importcfg_json["compilers"]:
+                    _importcfg[compiler["name"]] = compiler["options"]
+            except BaseException:
+                log.error("importcfg file failed validation")
+
+
+class Compiler(object):
+    """
+    Represents the behavior of a specific compiler, including:
+    - The number of passes it performs.
+    - Implicitly defined macros, which may be flag-dependent.
+    """
+
+    def __init__(self, args):
+        self.name = os.path.basename(args[0])
+        self.args = args
+        self.passes = set(["default"])
+
+        # Check for any user-defined compiler behavior.
+        # Currently, users can only override default defines.
+        if _importcfg is None:
+            load_importcfg()
+        self.defines = extract_defines(_importcfg[self.name])
+
+    def get_passes(self):
+        return self.passes.copy()
+
+    def get_defines(self, pass_):
+        return self.defines.copy()
+
+    def get_include_paths(self, pass_):
+        return []
+
+    def get_include_files(self, pass_):
+        return []
+
+    def has_implicit_behavior(self, pass_):
+        return (self.get_defines(pass_) or
+                self.get_include_paths(pass_) or
+                self.get_include_files(pass_))
+
+    def get_configuration(self, pass_):
+        defines = self.get_defines(pass_)
+        include_paths = self.get_include_paths(pass_)
+        include_files = self.get_include_files(pass_)
+        return {"defines": defines,
+                "include_paths": include_paths,
+                "include_files": include_files}
+
+
+class ClangCompiler(Compiler):
+    """
+    Represents the behavior of Clang-based compilers.
+    """
+
+    device_passes = [
+        "spir64",
+        "x86_64",
+        "spir64_x86_64",
+        "spir64_gen",
+        "spir64_fpga"
+    ]
+
+    def __init__(self, args):
```

Comment:
FYI as a matter of codestyle, I personally find it really hard (because duck-typing) to swallow `args` assuming some iterable without getting unpacked. I think the association is that `args` is usually used like (`*args, **kwargs`) and unpacked, so I mentally assume multiple things will be passed into the function like so:

```python
def func(*args):
    for arg in args: ...
```

where I can use the function like `func(a, b, c, d)`,  whereas this current style I can only use it via `func((a, b, c, d))`.

Reply:
I think this is largely a case of a bad variable name, which I adopted here because of a historical convention within CBI.  I don't think we realized that `args` here could easily be confused with the Python `*args` convention.  These functions accept a list because that's what you get in each entry of a compilation database:

```json
    {
        "arguments": [
            "icpx",
            "-fsycl",
            "-fsycl-targets=spir64",
            "test.cpp"
        ],
        "directory": "/home/",
        "file": "test.cpp"
    }
```

If we changed to `*args`, wouldn't we need to unpack the compilation database entry to call the function?  Like:
```python
Compiler(*entry["arguments"])
```

Not having to remember to unpack the list seems easier for this use case.  How would you feel about us renaming the variable from `args` to something like `argv`?  When we (eventually) adopt static typing as well, it would look like this:

```python
def __init__(self, argv: list):
```

If you're okay with this, I'll open a new issue to keep track of it. I'd rather make this change for all the functions accepting `args` at the same time, and do it in a separate PR.

Reply:
Maybe `build_args: list[str]`?

+1 for new issue/separate PR